Task 1

MadMan1996 opened the pull request:
Реализовал логику простого сервиса поддержки.
Создал сервлет, добавил маппинг севлета, покрыл код тестами, задеплоил приложение в томкат, используя докер.

MadMan1996 committed:
Add maven dependencies for testing, servlet

MadMan1996 committed:
Create util class for storing app data

MadMan1996 committed:
Add test for support servlet

MadMan1996 committed:
Add mapping for support servlet

MadMan1996 committed:
Set up tomcat with deployed servlet by docker-compose

MadMan1996 committed:
Add project description

MadMan1996 committed:
gitignore settings

MadMan1996 committed:
Add surefire-plugin to run test when build
Add encoding properties for project

MadMan1996 committed:
Fix readme style

rockstarhamster commented on .gitignore:
```diff
@@ -0,0 +1,38 @@
+!**/src/main/**/target/
+!**/src/test/**/target/
+
+### IntelliJ IDEA ###
```
Это файл для локальных нужд не следует его тянуть в репос. Например в нем помечаются для игнорирования файлы, которые inelij сгенерировала для Вашего удобства

MadMan1996 replied:
Пришли к выводу, что данный файл необходимо версионировать.

rockstarhamster commented on src/main/java/util/DataStore.java:
```diff
@@ -0,0 +1,27 @@
+import java.util.Set;
+
+public class DataStore {
+    private static final Set<String> SUPPORT_PHRASES = new HashSet<>();
```
Подумайте над доступом из нескольких потоков (много юзеров одновременно используют ваш API)

MadMan1996 replied:
Заменил реализаци на ConcurrentHashMap

rockstarhamster commented on src/test/java/servlets/SupportServletTest.java:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    @Test
+    public void doGetTest() throws ServletException, IOException {
```
В тестах всегж лучше завязываться на явные литералы. Например в этом тесте замокать обхект который выдает строку и удостовериться, что именно эта строка была записана в reponse

MadMan1996 replied:
Исправил в соответствии с комментариями. Замокал DataStore.getInstance() и метод обьекта DataStore, который возвращает фразу, для явного сравнения литералов

rockstarhamster commented on src/main/java/util/DataStore.java:
```diff
@@ -0,0 +1,27 @@
+import java.util.Random;
+import java.util.Set;
+
+public class DataStore {
```
Такую реализацию Вам будет достаточно проблематично тестировать, хотя и вполне возможно

MadMan1996 replied:
Обновил DataStore по принципу Singletone. Перенес статические методы в обькт

rockstarhamster commented on src/main/java/util/DataStore.java:
```diff
@@ -0,0 +1,27 @@
+    public static void addSupportPhrase(String phrase) {
+        SUPPORT_PHRASES.add(phrase);
+    }
+    public static Set<String> getSupportPhrases(){
```
Этот метод сильно нарушает инкапсуляцию. Зачем вообще тогда было создавать DataStore если так просто можно повлиять на его внутреннее состояние?

rockstarhamster commented on src/test/java/servlets/SupportServletTest.java:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    @Test
+    public void doPostWithWrongContentTypeTest() throws ServletException, IOException {
```
Молодец, что написали тесты на негативные сценарии!

rockstarhamster commented:
Очень качественно выполненное домашнее задание, Вы большой молодец!

MadMan1996 committed:
Сделал обьект по паттерну Singleton. Перенес статические методы в методы обьекта.
Обновил метод getSupportPhrases() - возврящает копию состояния, чтобы не нарушать инкапсуляцию.

MadMan1996 committed:
Изменил зависимость mockito-core to mockito-inline, чтобы мокировать статические методы.

MadMan1996 committed:
DataSoure - singleton

MadMan1996 committed:
Обновил doGetTest. Создал мок для DataSource, для явного сравнения литералов.
